src/main/java/udpapi2/UdpApi.java

```diff
@@ -0,0 +1,341 @@
+            aniDbIp = InetAddress.getByName(configuration.getAnidbHost());
+            aniDbPort = configuration.getAnidbPort();
+        } catch (SocketException e) {
+            logger.severe(STR."Failed to create socket \{e.getMessage()}");
+            return false;
+        } catch (UnknownHostException e) {
+            logger.severe(STR."Failed to resolve host \{e.getMessage()}");
+            return false;
+        }
+        receiveFuture = executorService.submit(new Receive(this));
+        isInitialized = true;
+        return true;
+    }
+
+
+    public void queueCommand(Command command) {
+        commandQueue.add(command);
+        if (!isSendScheduled && loginStatus != LoginStatus.LOGIN_PENDING) {
+            scheduleNextCommand();
+        }
+    }
+
+    private boolean queueLogin() {
+        if (!isInitialized) {
+            logger.warning("Must be initialized before logging in");
+            return false;
+        }
+        if (isSendScheduled) {
+            logger.warning("Command is scheduled, not scheduling login");
+            return false;
+        }
+        if (loginStatus == LoginStatus.LOGIN_PENDING) {
+            logger.warning("Login already scheduled, not scheduling login");
+            return false;
+        }
+        if (loginStatus == LoginStatus.LOGGED_IN) {
+            logger.warning("Already logged in, not scheduling login");
+            return false;
+        }
+        try {
+            val command = LoginCommand.Create(username, password);
+            scheduleCommand(command, getNextSendDelay());
+            isSendScheduled = true;
+            loginStatus = LoginStatus.LOGIN_PENDING;
+            return true;
+        } catch (IllegalArgumentException e) {
+            logger.warning(STR."Username or password is empty: \{e.getMessage()}");
+            return false;
+        }
+    }
+
+    private void scheduleNextCommand() {
+        if (isSendScheduled) {
+            return;
+        }
+        if (shouldWaitLong) {
+            queueLogin();
+            return;
+        }
+        val command = commandQueue.poll();
+        if (command == null) {
+            if (queries.isEmpty()) {
+                QueryId.reset();
+            }
+            if (loginStatus == LoginStatus.LOGGED_IN) {
+                queueLogout(false);
+            }
+            return;
+        }
+        if (logoutFuture != null) {
+            logoutFuture.cancel(false);
+            logoutFuture = null;
+        }
+        if (command.isNeedsLogin() && loginStatus != LoginStatus.LOGGED_IN) {
+            if (loginStatus == LoginStatus.LOGGED_OUT) {
+                queueLogin();
+            }
+            queueCommand(command);
+            return;
+        }
+        scheduleCommand(command, getNextSendDelay());
+    }
+
+    private void scheduleCommand(@NotNull Command command, Duration delay) {
+        logger.info(STR."Scheduling command \{command.toString()} in \{delay.toMillis()} ms");
```

Comment:
Might I suggest sanitizing the command output prior to logging it to mask the user/pass fields?
While the following code worked for me, I do not claim it is a good way to do this.
```
    private String sanitizedLogString(String message) {
        String sanitizedQueryString = message.replaceAll("(pass=)[^,}]*", "$1****")
                .replaceAll("(user=)[^,}]*", "$1****");
        return sanitizedQueryString;
    }
```
```suggestion
       logger.info(STR."Scheduling command \{sanitizedLogString(command.toString())} in \{delay.toMillis()} ms");
```

Reply:
oh woops.... that's a very good point 🙈 i'll implement that before merging

Reply:
i've added a todo for now, will fix this week